### PR TITLE
QC-479: merging supports vectors of TObject

### PR DESCRIPTION
**Pull request by justonedev1**
the first version of requested task, just to check whether I am doing correct thing before adding tests

### CODE REVIEW

File: Utilities/Mergers/include/Mergers/ObjectStore.h [hunk 1]
```diff
@@ -19,7 +19,9 @@
 
 #include <variant>
 #include <memory>
-#include "Framework/DataRef.h"
+#include <vector>
+#include <Framework/DataRef.h>
+#include <Framework/DataAllocator.h>
```
Comment (knopers8) [inline Utilities/Mergers/include/Mergers/ObjectStore.h]: forward declare

File: Utilities/Mergers/include/Mergers/ObjectStore.h [hunk 2]
```diff
@@ -19,7 +19,9 @@
 
 #include <variant>
 #include <memory>
-#include "Framework/DataRef.h"
+#include <vector>
+#include <Framework/DataRef.h>
```
Comment (knopers8) [inline Utilities/Mergers/include/Mergers/ObjectStore.h]: not needed?

File: Utilities/Mergers/src/ObjectStore.cxx [hunk 3]
```diff
@@ -18,14 +18,63 @@
...
+template <typename... Args>
+static std::string concat(Args&&... arguments)
+{
+  std::ostringstream ss;
+  (ss << ... << arguments);
+  return std::move(ss.str());
+}
+
+template <typename TypeToRead>
+void* readObject(TypeToRead&& type, o2::framework::FairTMessage& ftm)
```
Comment (knopers8) [inline Utilities/Mergers/src/ObjectStore.cxx]: Why is it a template? Doesn't `ftm.ReadObjectAny` expect a `TClass*` anyway?

File: Utilities/Mergers/include/Mergers/ObjectStore.h [hunk 4]
```diff
@@ -29,17 +31,27 @@ namespace o2::mergers
 class MergeInterface;
 
 using TObjectPtr = std::shared_ptr<TObject>;
+using VectorOfTObject = std::vector<TObject*>;
+using VectorOfTObjectPtr = std::vector<TObjectPtr>;
 using MergeInterfacePtr = std::shared_ptr<MergeInterface>;
```
Comment (knopers8) [inline Utilities/Mergers/include/Mergers/ObjectStore.h]:
```suggestion
using TObjectShrPtr = std::shared_ptr<TObject>;
using VectorOfTObjectPtr = std::vector<TObject*>;
using VectorOfTObjectShrPtr = std::vector<TObjectPtr>;
using MergeInterfaceShrPtr = std::shared_ptr<MergeInterface>;
```
I think it is confusing later in the code that `VectorOfTObject` stores pointers while `VectorOfTObjectPtr` stores shared pointers. Perhaps we could make it more explicit.

File: Utilities/Mergers/include/Mergers/ObjectStore.h [hunk 5]
```diff
@@ -29,17 +31,27 @@ namespace o2::mergers
...
+using ObjectStore = std::variant<std::monostate, TObjectPtr, VectorOfTObjectPtr, MergeInterfacePtr>;
 
 namespace object_store_helpers
 {
 
 /// \brief Takes a DataRef, deserializes it (if type is supported) and puts into an ObjectStore
 ObjectStore extractObjectFrom(const framework::DataRef& ref);
 
+/// \brief Helper function that converts vector of smart pointers to the vector of raw pointers that is serializable
+///        Destroy original vector after destroying vector with raw pointers to avoid undefined behavior
```
Comment (knopers8) [inline Utilities/Mergers/include/Mergers/ObjectStore.h]: Why does destroying the raw pointer vector invalidate the original vector? If you meant "the original vector should live longer than the observer vector", please reformulate accordingly.

File: Utilities/Mergers/include/Mergers/ObjectStore.h [hunk 6]
```diff
@@ -29,17 +31,27 @@ namespace o2::mergers
...
+VectorOfTObject toRawPointers(const VectorOfTObjectPtr&);
```
Comment (knopers8) [inline Utilities/Mergers/include/Mergers/ObjectStore.h]:
```suggestion
VectorOfTObject toRawObserverPointers(const VectorOfTObjectPtr&);
```
perhaps change the name as such to be clearer about the fact that the returned vector does not own the pointers.

File: Utilities/Mergers/src/ObjectStore.cxx [hunk 7]
```diff
@@ -34,49 +83,89 @@ ObjectStore extractObjectFrom(const framework::DataRef& ref)
...
+{
+  // NOTE: MT - it might be worth it to create custom stack allocators for this case
+  VectorOfTObject result{};
+  result.reserve(vector.size());
+  std::transform(vector.begin(), vector.end(), std::back_inserter(result), [](const auto& ptr) { return ptr.get(); });
+  return result;
+}
+
+template <typename TypeToSnapshot>
+struct snapshoter {
```
Comment (knopers8) [inline Utilities/Mergers/src/ObjectStore.cxx]:
```suggestion
struct Snapshoter {
```
We use UpperCamelCase for structures and classes.

File: Utilities/Mergers/include/Mergers/ObjectStore.h [hunk 8]
```diff
@@ -29,17 +31,27 @@ namespace o2::mergers
...
+
+/// \brief Used in FullHistorMerger's and IntegratingMerger's publish function. Checks mergedObject for every state that is NOT monostate
+///        and creates snapshot if underlying object to the framework
```
Comment (knopers8) [inline Utilities/Mergers/include/Mergers/ObjectStore.h]:
```suggestion
///        and creates snapshot of underlying object to the framework
```

Comment (knopers8) [inline Utilities/Mergers/include/Mergers/ObjectStore.h]: please document that it returns `true` if the provided object could be snapshot.

File: Utilities/Mergers/src/MergerAlgorithm.cxx [hunk 9]
```diff
@@ -128,6 +130,20 @@ void merge(TObject* const target, TObject* const other)
   }
 }
 
+void merge(VectorOfTObjectPtr& targets, const VectorOfTObjectPtr& others)
+{
+  for (const auto& other : others) {
+    if (const auto target_same_name = std::find_if(targets.begin(), targets.end(), [&other](const auto& target) {
```
Comment (knopers8) [inline Utilities/Mergers/src/MergerAlgorithm.cxx]: we use lowerCamelCase for local variables (see https://github.com/AliceO2Group/CodingGuidelines)

File: Utilities/Mergers/src/MergerAlgorithm.cxx [hunk 10]
```diff
@@ -128,6 +130,20 @@ void merge(TObject* const target, TObject* const other)
...
+          return std::string_view{other->GetName()} == std::string_view{target->GetName()};
+        });
+        target_same_name != targets.end()) {
+      merge(target_same_name->get(), other.get());
+    } else {
+      targets.push_back(other);
```
Comment (knopers8) [inline Utilities/Mergers/src/MergerAlgorithm.cxx]: I am wondering if it wouldn't be better to be consistent with [merging TCollections](https://github.com/AliceO2Group/AliceO2/blob/e7d0c164ba90dd7bbfed4e5a9ca815c9ff7d8063/Utilities/Mergers/src/MergerAlgorithm.cxx#L80), i.e. if an object does not exist in the target, we clone it. It has an advantage that the `other` vector/collection is not modified nor shared with `target`, thus it is less misleading when testing and possibly in the other future calls in the source code.

File: Utilities/Mergers/src/MergerAlgorithm.cxx [hunk 11]
```diff
@@ -16,18 +16,20 @@
 
 #include "Mergers/MergerAlgorithm.h"
 
-#include "Mergers/MergeInterface.h"
 #include "Framework/Logger.h"
+#include "Mergers/MergeInterface.h"
```
Comment (knopers8) [inline Utilities/Mergers/src/MergerAlgorithm.cxx:20]:
```suggestion
```
Double include

File: Utilities/Mergers/src/mergersTopologyExample.cxx [hunk 12]
```diff
@@ -39,142 +41,123 @@ void customize(std::vector<CompletionPolicy>& policies)
 #include <memory>
 #include <random>
 
-using namespace std::chrono;
+using SubSpecificationType = o2::header::DataHeader::SubSpecificationType;
 
-// clang-format off
```
Comment (knopers8) [inline Utilities/Mergers/src/mergersTopologyExample.cxx:44]: I am glad to see that now clang-format behaves better with lambdas in structures than it used to :)

File: Utilities/Mergers/test/test_ObjectStore.cxx [hunk 13]
```diff
@@ -29,89 +30,114 @@ using namespace o2::framework;
 using namespace o2::mergers;
 
 // Simple test to do root deserialization.
-BOOST_AUTO_TEST_CASE(TestObjectExtraction)
+BOOST_AUTO_TEST_SUITE(TestObjectExtraction)
+
+template <typename TypeToDataRef>
+
```
Comment (knopers8) [inline Utilities/Mergers/test/test_ObjectStore.cxx]:
```suggestion
```

File: Utilities/Mergers/test/test_ObjectStore.cxx [hunk 14]
```diff
@@ -9,6 +9,7 @@
 // granted to it by virtue of its status as an Intergovernmental Organization
 // or submit itself to any jurisdiction.
 
+#include <TH1.h>
```
Comment (knopers8) [inline Utilities/Mergers/test/test_ObjectStore.cxx:12]: why is it needed here? I would rather add `<TH1F.h>` next to the other ROOT histogram include, since you used that one in the new test case.

File: Utilities/Mergers/test/test_ObjectStore.cxx [hunk 15]
```diff
@@ -29,89 +30,114 @@ using namespace o2::framework;
...
+  DataRef ref = makeDataRef(&vectorToDataRef);
+  auto objStore = object_store_helpers::extractObjectFrom(ref);
+  BOOST_CHECK(std::holds_alternative<VectorOfTObjectPtr>(objStore));
+  auto extractedVector = std::get<VectorOfTObjectPtr>(objStore);
+  BOOST_CHECK(extractedVector.size() == 1);
+  auto* extractedHisto = dynamic_cast<TH1F*>(extractedVector[0].get());
+  BOOST_CHECK(gsl::span(histo->GetArray(), histo->GetSize()) == gsl::span(extractedHisto->GetArray(), extractedHisto->GetSize()));
+}
+
+BOOST_AUTO_TEST_SUITE_END()
```
Comment (knopers8) [inline Utilities/Mergers/test/test_ObjectStore.cxx:142]: Thanks for refactoring the file to use SUITE and CASE. I was not aware of the possible distinction, indeed it is better this way. However, next time please make such improvements as separate commits or PRs, since it obfuscates the commit history.

File: Utilities/Mergers/include/Mergers/LinkDef.h [hunk 16]
```diff
@@ -18,5 +18,6 @@
 #pragma link C++ class o2::mergers::MergeInterface + ;
 #pragma link C++ class o2::mergers::CustomMergeableObject + ;
 #pragma link C++ class o2::mergers::CustomMergeableTObject + ;
+#pragma link C++ class std::vector < TObject*> + ;
```
Comment (knopers8) [inline Utilities/Mergers/include/Mergers/LinkDef.h:21]:
```suggestion
#pragma link C++ class std::vector < TObject* > + ;
```

**Review by knopers8 — CHANGES_REQUESTED** [review]
Thank you for this PR, I think it is very good. I have quite a lot of suggestions, some only pedantic, but I do not see anything fundamentally wrong.

File: Utilities/Mergers/include/Mergers/LinkDef.h (hunk 16 above)
Comment (justonedev1) [reply Utilities/Mergers/include/Mergers/LinkDef.h:21]: this is clang formatted (surprisingly enough)

File: Utilities/Mergers/test/test_ObjectStore.cxx (hunk 15 above)
Comment (justonedev1) [reply Utilities/Mergers/test/test_ObjectStore.cxx:142]: I thought that this refactoring of test suite, I was working on, was okay to be in this PR. But I will remember that for the future.

File: Utilities/Mergers/src/mergersTopologyExample.cxx (hunk 12 above)
Comment (justonedev1) [reply Utilities/Mergers/src/mergersTopologyExample.cxx:44]: it behaved weirdly because of c-style casts of lambdas, when I changed that to static_casts it started to behave correctly

File: Utilities/Mergers/src/MergerAlgorithm.cxx (hunk 10 above)
Comment (justonedev1) [reply Utilities/Mergers/src/MergerAlgorithm.cxx]: That is not that bad of an idea, However I am really not sure, how to do it properly, as Clone method on TObject requires new name and I am not sure which name should I give it

File: Utilities/Mergers/include/Mergers/LinkDef.h (hunk 16 above)
Comment (knopers8) [reply Utilities/Mergers/include/Mergers/LinkDef.h:21]: ok, then let's leave it happy

File: Utilities/Mergers/src/MergerAlgorithm.cxx (hunk 10 above)
Comment (knopers8) [reply Utilities/Mergers/src/MergerAlgorithm.cxx]: I don't think ROOT is going to complain about duplicate names if you `Clone`, at least I did not see that due to the use of `Clone` when merging `TCollections`. Does it produce a warning in your case?

Comment (justonedev1) [reply Utilities/Mergers/src/MergerAlgorithm.cxx]: You are right, everything is fine, just by calloing `Clone()`

**Review by knopers8 — APPROVED** [review]

**Comment by knopers8** [comment]
Thank you!
